### PR TITLE
[AutoDiff upstream] Add `@noDerivative` flag to `SILParameterInfo`.

### PR DESCRIPTION
**Pull request by dan-zheng**
The `@noDerivative` attribute marks the non-differentiability parameters of a
`@differentiable` function type. All parameters except those marked with
`@noDerivative` are differentiability parameters.

For example, `@differentiable (Float, @noDerivative Float) -> Float` is only
differentiable with respect to its first parameter.

The `@noDerivative` attribute is represented as a
`SILParameterDifferentiability` bit on `SILParameterInfo`.

The existing `@noDerivative` AST attribute is now lowered to SIL.

Add round-trip serialization and SILGen tests.

Resolves TF-872.

### CODE REVIEW

**Comment by dan-zheng** [comment]
@swift-ci Please test

**Comment by dan-zheng** [comment]
@jckarter: would you like to review this PR, or suggest a reviewer? I'm asking you since you seem like a SIL code owner.

This PR is part of our effort to [upstream differentiable programming](https://bugs.swift.org/browse/TF-820). It lowers [an existing AST flag](https://github.com/apple/swift/pull/28278) to SIL and should be relatively non-controversial.

If there's no activity, I'll merge within a day to unblock progress. Happy to address feedback at any time!

**Review by rxwei — APPROVED** [review]

**Comment by jckarter** [comment]
Looks good to me.

**Pull request merged**
